crates/bevy_animation/src/lib.rs

```diff
@@ -65,6 +65,43 @@ pub struct VariableCurve {
     pub interpolation: Interpolation,
 }
 
+impl VariableCurve {
+    /// Find the index of the keyframe at or before the current time.
+    ///
+    /// Returns [`None`] if the curve is finished or not yet started.
+    pub fn find_current_keyframe(&self, seek_time: f32) -> Option<usize> {
+        // An Ok(keyframe_index) result means an exact result was found by binary search
+        // An Err result means the keyframe was not found, and the index is the keyframe
+        // PERF: finding the current keyframe can be optimised
+        let search_result = self
+            .keyframe_timestamps
+            .binary_search_by(|probe| probe.partial_cmp(&seek_time).unwrap());
+
+        // We want to find the index of the keyframe before the current time
+        // If the keyframe is past the second-to-last keyframe, the animation cannot be interpolated.
+        let step_start = match search_result {
+            // An exact match was found, and it is the last or second last keyframe.
+            // This means that the curve is finished.
+            Ok(n) if n >= self.keyframe_timestamps.len() - 1 => return None,
+            // An exact match was found, and it is not the last or second last keyframe.
+            Ok(i) => i,
+            // No exact match was found, and the curve is not yet started.
+            // This occurs because the binary search returns the index of where we could insert a value
+            // without disrupting the order of the vector.
+            // If the value is less than the first element, the index will be 0.
+            Err(0) => return None,
+            // No exact match was found, and the curve is finished.
+            Err(n) if n > self.keyframe_timestamps.len() - 1 => return None,
```

Comment:
This would be clearer if this was >= the length, but I wanted to leave this untouched until the comment above was resolved.